### ISSUE TITLE
Lab 3 - Review by MattColu

### ISSUE THREAD

**Issue by MattColu** [gabrjc/Computational-Intelligence-2022#6]
# Nim

## 3.1

I like the inclusion of both the brute force and nimsum approach. The cost-minimizing part in the "back to safety" function is also a good idea.

## 3.2

This is the same approach that I followed, and I think it's the one that best takes advantage of the GA paradigm.

I would suggest moving the checking of `LAST_CONDITION` and the management of the amount (`"take_one"`, `"take_all"`, `"take_all_except_one"`, etc.) in their own functions, since the same code is repeated for each column policy.

## 3.3

As you said, some sort of pruning would be quite beneficial, nothing much to add on the algorithm itself.

## 3.4

As well, not much to add on the algorithm. It's a solid implementation of Q-Learning (I appreciated all the comments sprinkled throughout). Training the agent against itself is also a great idea, but you might get some skewed results (the Professor said).

---

In conclusion, all the algorithms are pretty solid, although some still improvable (especially on the efficiency front).

Another thing I would suggest is splitting the first and second task and making the overall code a bit tidier (for example specifying new functions in the notebooks while importing the recurring ones from a support file), it does get a bit lengthy at times.

Although not the prettiest I can see that a lot of work went into this assignment, well done! 👍
